Fix: message-id in postprocessor/gelf-chunking

BharatKJain opened the pull request:
# Pull request

## Description

<!-- please add a description of what the goal of this pull request is -->

Changed message-id from auto-increment ID to randomized ID in postprocessor/gelf_chunking.rs

HELP NEEDED: I have avoided adding hostname while producing message-id, I am not sure how can we handle adding hostname, please suggest.

## Related

<!-- please include links to related issues are RFCs, remove if not required  -->

- [Issue-820](https://github.com/tremor-rs/tremor-runtime/issues/820)
- [Reference code](https://github.com/osiegmar/logback-gelf/blob/master/src/main/java/de/siegmar/logbackgelf/MessageIdSupplier.java#L61)

## Checklist

<!--
Please fill out the checklist below.

If an RFC is required and not submitted yet the PR will be tagged as RFC required and blocked
until the RFC is submitted and approved.

As a rule of thumb, bugfixes or minimal additions that have no backwards impact and are fully
self-contained usually do not require an RFC. Larger changes, changes to behavior, breaking changes
usually do. If in doubt, please open a ticket for a PR first to discuss the issue.

-->

* [ ] The RFC, if required, has been submitted and approved
* [ ] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [ ] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [ ] The performance impact of the change is measured (see below)

## Performance

<!--
Measure or reason about the performance impact of the change.

A rough indication is sufficient here, we often use the `real-world-json-throughput` example to

1. cargo build -p tremor-cli --release (on main)
2. ./bench/run real-workflow-throughput-json
3. repeat on main

Share the two throughput numbers and the benchmark that produced them.

NOTE: We are fully aware this is not a perfect method, but it is a tradeoff between preventing large
performance degradation and putting an unreasonable burden on contributors.

NOTE: the total number is irrelevant as it will vary from system to system. The delta is what
matters.

If the benchmarks fail on your system, note it in the issue, and someone will try to run them for
you.

If your changes do not affect performance, and you can argue this, do it in this section, it is a
valid response.

-->

Licenser reviewed (approved):
👍  looks reasonable nothing to prevent using this I see, great documentation too :)

two things I notice.

1) it would be nice to mention how message id's are generated in the docs (the `//! ` section)

2) if you are up for a challenge: we try to keep all time and randomness out of tremor to allow for deterministic replays. We do this by using `ingest_ns` for as a random seed, and for times, that way a event that is logged with it's ingest ns can be replayed and generate the exact same out put. [The random function](https://github.com/tremor-rs/tremor-runtime/blob/main/tremor-script/src/std_lib/random.rs#L38) is a good example. It'd be interesting to see this same concept re-used for this to allow repeatable yet still random message id's; one way would be to use ingest-ns instead of the current epoch (which would be nice anyway as looking uo time isn't fast), and then seed the RNG somehow (probably not with the ingest ns as that would make it useless) but perhaps with the fist n bytes of the message? or with some bytes of a hash of the message?

BharatKJain commented:
1. Okay, will do.

2. Trying to think-out-loud, so let's say if we create a hash of a message but repetitive logs will create same message-id which is a problem because message-id has to be unique in nature, collisions can cause problems when we're decoding the message on the server side. I am not completely sure but ideally we would want to have uniqueness in the message-id to make sure that we are not breaking the server GELF-decoding. 

(How it will break server-side due to collision? So message-id is a way of determining if the UDP packet is associated with already existing log or it's for a new log, when we're sending same message-id for multiple logs then server behaviour will be to merge the data-together which will end-up breaking the log)

TBH I am also trying to figure this out, please share any suggestions, am I thinking right? 😅

Licenser commented:
Ja just the message content would not work, I'm still considering if message content + ingest_ns (nanosecond when the message was registered at tremor) would be enough, if a server produces the same log twice in the same nanosecond that'd be very odd (but not impossible) OTOH having two random generated numbers be the same is also odd (but not impossible) it would also one a more deterministic failure case "When messages with the same content arrive at exactly the same time they will get duplicated message ids" instead of "if the RNG hates you, you'll get duplicated message ids"

Licenser commented:
Sorry for all the forth and back, so I've been thinking about a good way to make this:

1) fast
2) deterministic
3) non-coliding

and wanted to throw out a suggestion using the ingest_ns + an incremental id as a message ID. This is:

1) fast since we just have to look at two integers, no RNG, no hashing
2) deterministic since the ingest_ns is settable, and the incremental counter is well, incremental so determinstic as well
3) it avoids duplicates by not creating duplicates on the same system due to incremtal id's and makes them extremely unlikely on multiple systems due to nano second timestamps involved.

BharatKJain commented:
Hmm, I have changed the code to this:
```
(epoch_timestamp & BITS_13) | (data_sha_hash_u64 & !BITS_13) | (thread_id_u64 & BITS_13)
```

epoch_timestamp is `ingest_ns` for process function, for finish function I am taking the current timestamp as there's no `ingest_ns` in the downstream function call.

1. Does the above mentioned approach look good?
2. I didn't understand why are they using `& BITS_13` in the java library, I think if we flip the bits using OR `|`, that should suffice - is that right?
3. I have also added a unit-test as well as removed id from the struct for Gelf - I didn't understand if there's a need to add id in that struct, I cannot find code where we are preserving any message-id that's coming from downstream, please check this as well.

BharatKJain commented:
> Sorry for all the forth and back, so I've been thinking about a good way to make this:
> 
>     1. fast
> 
>     2. deterministic
> 
>     3. non-coliding
> 
> 
> and wanted to throw out a suggestion using the ingest_ns + an incremental id as a message ID. This is:
> 
>     1. fast since we just have to look at two integers, no RNG, no hashing
> 
>     2. deterministic since the ingest_ns is settable, and the incremental counter is well, incremental so determinstic as well
> 
>     3. it avoids duplicates by not creating duplicates on the same system due to incremtal id's and makes them extremely unlikely on multiple systems due to nano second timestamps involved.

Can I keep thread-id just to reduce the collision probability more? 😅

BharatKJain commented:
Okay, I have made the changes as suggested in the latest comment.

```
(epoch_timestamp & BITS_13) | (auto_increment_id & !BITS_13) | (thread_id_u64 & BITS_13)
```

- Added a unit-test for it.
- Renamed id to auto_increment_id


Let me know your thoughts!

Licenser commented:
I would remove the thread ID. It is breaking the distribution and making the date less unique also it prevents event id's to be replayable. Basiclly it will result in the lower 13 bit to ave 3 times as many 1's as 0's making them more likely to collide while also making the id not determinstic.

The second problem I spot is that by removing the lower 13 bit from increment, that part will have no effect for the first 8192 messages and then only change every 8192 messages. My suggestion would be to shift it by 13 bits instead of truncating them.
 
Lastly I'd probably pull in a few more bits form the timestap, 13 bit are only `0,0000082 secs` even if you bump it to 16 that means you get a window of `0,000066 secs` + a 48bit counter (or 281.474.976.710.656 events to count)
 
You'd end up with something like this:


```rust
(epoch_timestamp & 0xFF_FF) | (auto_increment_id << 16 )
 ``
 
 do you think that would solve the original problem?

Licenser reviewed (approved):
wooh wooh

BharatKJain commented:
I will fix the clippy checks and DCO, please allow me sometime.


---

Honestly I didn't anticipate this to happen but I also had built otel gelf-exporter. :)

[Opentelemetry PR for gelf-exporter (WIP)](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35820)

BharatKJain commented:
Line 190:

```
let current_epoch_timestamp = u64::try_from(SystemTime::now().duration_since(UNIX_EPOCH).expect("SystemTime before UNIX EPOCH!").as_nanos())?;
```

Not sure how to handle this better, is this okay?

Licenser commented:
We'd want to avoid the expect. Let me explain why.

`expect` and `unwrap` cause a crash, in some applications that's okay since, say you have a CLI if something goes wrong you might just want to error out and restart from scratch. Now tremor is a bit more complex there might be more then one pipeline running so we don't want one pipeline affect the progress of another. If we'd crash in pipeline A we'd take pipeline B down with it - that's not desirable. So we handle errors and let them bubble up so only the pipeline causing the issue is affected.

To do that there are a few ways `?` in many places work, but if the error type can't be mapped something like `map_err(...)?` is a good alternative.

BharatKJain commented:
@Licenser @darach Does it look good now?

(I have fixed the DCO, format, clippy-check, code quality checks)

darach reviewed (approved):
LGTM 🚀
